File: problems/103/103.binary-tree-zigzag-level-order-traversal.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=103 lang=java
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Deque<TreeNode> q = new ArrayDeque<>();
        int level = 0;

        if (root != null) q.offerFirst(root);

        while (!q.isEmpty()) {
            List<Integer> list = new ArrayList<>();
            int size = q.size();
            for (int i = 0; i < size; i++) {
                if(level % 2 == 0) {
                    TreeNode cur = q.pollLast();
                    list.add(cur.val);
                    if(cur.left != null) q.offerFirst(cur.left);
                    if(cur.right != null) q.offerFirst(cur.right);
                } else {
                    TreeNode cur = q.pollFirst();
                    list.add(cur.val);
                    if(cur.right != null) q.offerLast(cur.right);
                    if(cur.left != null) q.offerLast(cur.left);
                }
            }
            res.add(list);
            level++;
        }

        return res;
    }
}
// @lc code=end

File: problems/103/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # level order with stack
        # flip between reverse true/false

        if not root:
            return []

        q = deque()
        q.append(root)
        res = []

        while q:
            curr = []
            for _ in range(len(q)):
                node = q.popleft()
                curr.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            
            res.append(curr)

        return list(curr if idx % 2 == 0 else reversed(curr) for idx, curr in enumerate(res))
        
        
# @lc code=end

